ICE `Illegal placeholder constraint blamed; should have redirected to other region relation`

<!--
[31mICE[0m: Rustc ./FF60DB0AA2A5457B323D7F14F3D440A121CF6A32F1462B314DDE91EA0FC02695.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_borrowck/src/region_infer/mod.rs:1767:9: 'Illegal placeholder constraint blamed; should have redirected to other region relation'', 'thread 'rustc' ($TID) panicked at compiler/rustc_borrowck/src/region_infer/mod.rs:1767:9: 'Illegal placeholder constraint blamed; should have redirected to other region relation''
File: /tmp/im/FF60DB0AA2A5457B323D7F14F3D440A121CF6A32F1462B314DDE91EA0FC02695.rs
-->

auto-reduced (treereduce-rust):
````rust
#![feature(inherent_associated_types)]

struct Foo<T>(T);

impl<'a> Foo<fn(&())> {
    type Assoc = &'a ();
}

fn main(_: for<'a> fn(Foo<fn(&'a ())>::Assoc)) {}

````

original:
````rust
#![feature(inherent_associated_types)]
#![feature(inherent_associated_types)]

// Check that we don't crash when printing inherent projections in diagnostics.

struct Foo<T>(T);

impl<'a> Foo<fn(& ())> {
    type Assoc = &'a ();
}

fn main(_: for<'a> fn(Foo<fn(&'a ())>::Assoc)) {} //~ ERROR `main` function has wrong type

````

Version information
````
rustc 1.91.0-nightly (8e2ed71ef 2025-09-12)
binary: rustc
commit-hash: 8e2ed71effd5f81bff319c0c7adaca42084e2a71
commit-date: 2025-09-12
host: x86_64-unknown-linux-gnu
release: 1.91.0-nightly
LLVM version: 21.1.1

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/8e2ed71effd5f81bff319c0c7adaca42084e2a71/compiler/rustc_borrowck/src/region_infer/mod.rs#L1761-L1773

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
 --> /tmp/icemaker_global_tempdir.XMXxmSTCf6Ug/rustc_testrunner_tmpdir_reporting.qrZtQYvCdou4/mvce.rs:1:12
  |
1 | #![feature(inherent_associated_types)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0207]: the lifetime parameter `'a` is not constrained by the impl trait, self type, or predicates
 --> /tmp/icemaker_global_tempdir.XMXxmSTCf6Ug/rustc_testrunner_tmpdir_reporting.qrZtQYvCdou4/mvce.rs:5:6
  |
5 | impl<'a> Foo<fn(&())> {
  |      ^^ unconstrained lifetime parameter

error[E0308]: mismatched types
 --> /tmp/icemaker_global_tempdir.XMXxmSTCf6Ug/rustc_testrunner_tmpdir_reporting.qrZtQYvCdou4/mvce.rs:9:12
  |
9 | fn main(_: for<'a> fn(Foo<fn(&'a ())>::Assoc)) {}
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
  |
  = note: expected struct `Foo<for<'a> fn(&'a ())>`
             found struct `Foo<for<'a> fn(&'a ())>`

error[E0580]: `main` function has wrong type
 --> /tmp/icemaker_global_tempdir.XMXxmSTCf6Ug/rustc_testrunner_tmpdir_reporting.qrZtQYvCdou4/mvce.rs:9:1
  |
9 | fn main(_: for<'a> fn(Foo<fn(&'a ())>::Assoc)) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ incorrect number of function parameters
  |
  = note: expected signature `fn()`
             found signature `fn(fn(&()))`


thread 'rustc' (1111115) panicked at compiler/rustc_borrowck/src/region_infer/mod.rs:1767:9:
Illegal placeholder constraint blamed; should have redirected to other region relation
stack backtrace:
   0:     0x7fbea2db1d93 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h69f84197283f92f2
   1:     0x7fbea3401b18 - core::fmt::write::hac51afc447b71987
   2:     0x7fbea2d66aa1 - std::io::Write::write_fmt::h7914652380e8da3b
   3:     0x7fbea2d77cc2 - std::sys::backtrace::BacktraceLock::print::he6531a518b7afaed
   4:     0x7fbea2d7dc09 - std::panicking::default_hook::{{closure}}::h4766ac8d31f0938d
   5:     0x7fbea2d7d733 - std::panicking::default_hook::hc214b0cee38d0644
   6:     0x7fbea1dc62b7 - std[1517c6b2b7ad33c0]::panicking::update_hook::<alloc[e8dcfcd713f5b5b9]::boxed::Box<rustc_driver_impl[f40c8d47e6531829]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7fbea2d7e02f - std::panicking::panic_with_hook::hac95cd1be6a69487
   8:     0x7fbea2d7ddb6 - std::panicking::panic_handler::{{closure}}::h71edd8dc49b8a93b
   9:     0x7fbea2d77df9 - std::sys::backtrace::__rust_end_short_backtrace::h028726c52dd94f61
  10:     0x7fbea2d584dd - __rustc[c3017cc57a982af5]::rust_begin_unwind
  11:     0x7fbe9f4c39f0 - core::panicking::panic_fmt::h9dca6dd2c2c3cf5b
  12:     0x7fbea42c0080 - <rustc_borrowck[cf86fedb87f30a8f]::region_infer::RegionInferenceContext>::best_blame_constraint
  13:     0x7fbea1b13481 - <rustc_borrowck[cf86fedb87f30a8f]::MirBorrowckCtxt>::report_region_errors
  14:     0x7fbe9fe5bd2c - rustc_borrowck[cf86fedb87f30a8f]::borrowck_check_region_constraints
  15:     0x7fbea465eab2 - <rustc_borrowck[cf86fedb87f30a8f]::root_cx::BorrowCheckRootCtxt>::do_mir_borrowck
  16:     0x7fbea46561d7 - rustc_borrowck[cf86fedb87f30a8f]::mir_borrowck
  17:     0x7fbea4655fb5 - rustc_query_impl[c1b7fe2d2380a2f3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c1b7fe2d2380a2f3]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4d64447825fbb932]::query::erase::Erased<[u8; 8usize]>>
  18:     0x7fbea35e35c5 - rustc_query_system[8bcbb518a438fd56]::query::plumbing::try_execute_query::<rustc_query_impl[c1b7fe2d2380a2f3]::DynamicConfig<rustc_data_structures[838be9b3682db453]::vec_cache::VecCache<rustc_span[c40e2ad170baf06]::def_id::LocalDefId, rustc_middle[4d64447825fbb932]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[8bcbb518a438fd56]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[c1b7fe2d2380a2f3]::plumbing::QueryCtxt, false>
  19:     0x7fbea35e2dc9 - rustc_query_impl[c1b7fe2d2380a2f3]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  20:     0x7fbea35ede42 - <rustc_middle[4d64447825fbb932]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface[9c243299d9070ddc]::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
  21:     0x7fbea35eb6d9 - rustc_interface[9c243299d9070ddc]::passes::analysis
  22:     0x7fbea35eab8b - rustc_query_impl[c1b7fe2d2380a2f3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c1b7fe2d2380a2f3]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4d64447825fbb932]::query::erase::Erased<[u8; 0usize]>>
  23:     0x7fbea44075a0 - rustc_query_system[8bcbb518a438fd56]::query::plumbing::try_execute_query::<rustc_query_impl[c1b7fe2d2380a2f3]::DynamicConfig<rustc_query_system[8bcbb518a438fd56]::query::caches::SingleCache<rustc_middle[4d64447825fbb932]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[c1b7fe2d2380a2f3]::plumbing::QueryCtxt, false>
  24:     0x7fbea44071f6 - rustc_query_impl[c1b7fe2d2380a2f3]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  25:     0x7fbea474ec23 - rustc_interface[9c243299d9070ddc]::passes::create_and_enter_global_ctxt::<core[fbb2f6aea0caeac7]::option::Option<rustc_interface[9c243299d9070ddc]::queries::Linker>, rustc_driver_impl[f40c8d47e6531829]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  26:     0x7fbea468eebe - rustc_interface[9c243299d9070ddc]::interface::run_compiler::<(), rustc_driver_impl[f40c8d47e6531829]::run_compiler::{closure#0}>::{closure#1}
  27:     0x7fbea45f12f8 - std[1517c6b2b7ad33c0]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[9c243299d9070ddc]::util::run_in_thread_with_globals<rustc_interface[9c243299d9070ddc]::util::run_in_thread_pool_with_globals<rustc_interface[9c243299d9070ddc]::interface::run_compiler<(), rustc_driver_impl[f40c8d47e6531829]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  28:     0x7fbea45f0fc8 - <<std[1517c6b2b7ad33c0]::thread::Builder>::spawn_unchecked_<rustc_interface[9c243299d9070ddc]::util::run_in_thread_with_globals<rustc_interface[9c243299d9070ddc]::util::run_in_thread_pool_with_globals<rustc_interface[9c243299d9070ddc]::interface::run_compiler<(), rustc_driver_impl[f40c8d47e6531829]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[fbb2f6aea0caeac7]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  29:     0x7fbea45f728d - std::sys::thread::unix::Thread::new::thread_start::h7cbbbba756171d4a
  30:     0x7fbe9de969cb - <unknown>
  31:     0x7fbe9df1aa0c - <unknown>
  32:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.91.0-nightly (8e2ed71ef 2025-09-12) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_borrowck] borrow-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors; 1 warning emitted

Some errors have detailed explanations: E0207, E0308, E0580.
For more information about an error, try `rustc --explain E0207`.

```

</p>
</details>

<!--
query stack:
#0 [mir_borrowck] borrow-checking `main`
#1 [analysis] running analysis passes on this crate

-->
@rustbot label +F-inherent_associated_types +F-inherent_associated_types

--- a/compiler/rustc_borrowck/src/region_infer/mod.rs
+++ b/compiler/rustc_borrowck/src/region_infer/mod.rs
@@ -1726,9 +1726,10 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                 // `BoringNoLocation` constraints can point to user-written code, but are less
                 // specific, and are not used for relations that would make sense to blame.
                 ConstraintCategory::BoringNoLocation => 6,
-                // Do not blame internal constraints.
-                ConstraintCategory::OutlivesUnnameablePlaceholder(_) => 7,
-                ConstraintCategory::Internal => 8,
+                // Do not blame internal constraints if we can avoid it. Never blame
+                // the `'region: 'static` constraints introduced by placeholder outlives.
+                ConstraintCategory::Internal => 7,
+                ConstraintCategory::OutlivesUnnameablePlaceholder(_) => 8,
             };
 
             debug!("constraint {constraint:?} category: {category:?}, interest: {interest:?}");

--- a/compiler/rustc_borrowck/src/type_check/mod.rs
+++ b/compiler/rustc_borrowck/src/type_check/mod.rs
@@ -505,6 +505,7 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         let mut constraints = Default::default();
         let mut liveness_constraints =
             LivenessValues::without_specific_points(Rc::new(DenseLocationMap::new(promoted_body)));
+        let mut deferred_closure_requirements = Default::default();
 
         // Don't try to add borrow_region facts for the promoted MIR as they refer
         // to the wrong locations.
@@ -512,6 +513,7 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             mem::swap(this.polonius_facts, polonius_facts);
             mem::swap(&mut this.constraints.outlives_constraints, &mut constraints);
             mem::swap(&mut this.constraints.liveness_constraints, &mut liveness_constraints);
+            mem::swap(this.deferred_closure_requirements, &mut deferred_closure_requirements);
         };
 
         swap_constraints(self);
@@ -536,6 +538,17 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             }
             self.constraints.outlives_constraints.push(constraint)
         }
+
+        // If there are nested bodies in promoteds, we also need to update their
+        // location to something in the actually body, not the promoted.
+        //
+        // We don't update the constraint categories of the resulting constraints
+        // as returns in nested bodies are a proper return, even if that nested body
+        // is in a promoted.
+        for (closure_def_id, args, _locations) in deferred_closure_requirements {
+            self.deferred_closure_requirements.push((closure_def_id, args, locations));
+        }
+
         // If the region is live at least one location in the promoted MIR,
         // then add a liveness constraint to the main MIR for this region
         // at the location provided as an argument to this method

--- a/compiler/rustc_borrowck/src/type_check/mod.rs
+++ b/compiler/rustc_borrowck/src/type_check/mod.rs
@@ -540,7 +540,7 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         }
 
         // If there are nested bodies in promoteds, we also need to update their
-        // location to something in the actually body, not the promoted.
+        // location to something in the actual body, not the promoted.
         //
         // We don't update the constraint categories of the resulting constraints
         // as returns in nested bodies are a proper return, even if that nested body

--- a/tests/ui/associated-inherent-types/hr-do-not-blame-outlives-static-ice.rs
+++ b/tests/ui/associated-inherent-types/hr-do-not-blame-outlives-static-ice.rs
@@ -1,0 +1,17 @@
+//@ compile-flags: -Zdeduplicate-diagnostics=yes
+
+// Regression test for #146467.
+#![feature(inherent_associated_types)]
+//~^ WARN the feature `inherent_associated_types` is incomplete
+
+struct Foo<T>(T);
+
+impl<'a> Foo<fn(&())> {
+    //~^ ERROR the lifetime parameter `'a` is not constrained by the impl trait
+    type Assoc = &'a ();
+}
+
+fn foo(_: for<'a> fn(Foo<fn(&'a ())>::Assoc)) {}
+//~^ ERROR mismatched types
+//~| ERROR higher-ranked subtype error
+fn main() {}

--- a/tests/ui/associated-inherent-types/hr-do-not-blame-outlives-static-ice.stderr
+++ b/tests/ui/associated-inherent-types/hr-do-not-blame-outlives-static-ice.stderr
@@ -1,0 +1,34 @@
+warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
+  --> $DIR/hr-do-not-blame-outlives-static-ice.rs:4:12
+   |
+LL | #![feature(inherent_associated_types)]
+   |            ^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information
+   = note: `#[warn(incomplete_features)]` on by default
+
+error[E0207]: the lifetime parameter `'a` is not constrained by the impl trait, self type, or predicates
+  --> $DIR/hr-do-not-blame-outlives-static-ice.rs:9:6
+   |
+LL | impl<'a> Foo<fn(&())> {
+   |      ^^ unconstrained lifetime parameter
+
+error[E0308]: mismatched types
+  --> $DIR/hr-do-not-blame-outlives-static-ice.rs:14:11
+   |
+LL | fn foo(_: for<'a> fn(Foo<fn(&'a ())>::Assoc)) {}
+   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
+   |
+   = note: expected struct `Foo<for<'a> fn(&'a ())>`
+              found struct `Foo<for<'a> fn(&'a ())>`
+
+error: higher-ranked subtype error
+  --> $DIR/hr-do-not-blame-outlives-static-ice.rs:14:1
+   |
+LL | fn foo(_: for<'a> fn(Foo<fn(&'a ())>::Assoc)) {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 3 previous errors; 1 warning emitted
+
+Some errors have detailed explanations: E0207, E0308.
+For more information about an error, try `rustc --explain E0207`.

--- a/tests/ui/higher-ranked/do-not-blame-outlives-static-ice.rs
+++ b/tests/ui/higher-ranked/do-not-blame-outlives-static-ice.rs
@@ -1,0 +1,12 @@
+//@ compile-flags: -Zdeduplicate-diagnostics=yes
+
+// Regression test for #146467.
+trait Trait { type Assoc; }
+
+impl Trait for fn(&()) { type Assoc = (); }
+
+fn f(_: for<'a> fn(<fn(&'a ()) as Trait>::Assoc)) {}
+//~^ ERROR implementation of `Trait` is not general enough
+//~| ERROR higher-ranked subtype error
+
+fn main() {}

--- a/tests/ui/higher-ranked/do-not-blame-outlives-static-ice.stderr
+++ b/tests/ui/higher-ranked/do-not-blame-outlives-static-ice.stderr
@@ -1,0 +1,17 @@
+error: implementation of `Trait` is not general enough
+  --> $DIR/do-not-blame-outlives-static-ice.rs:8:9
+   |
+LL | fn f(_: for<'a> fn(<fn(&'a ()) as Trait>::Assoc)) {}
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Trait` is not general enough
+   |
+   = note: `for<'a> fn(&'a ())` must implement `Trait`, for any lifetime `'0`...
+   = note: ...but `Trait` is actually implemented for the type `for<'a> fn(&'a ())`
+
+error: higher-ranked subtype error
+  --> $DIR/do-not-blame-outlives-static-ice.rs:8:1
+   |
+LL | fn f(_: for<'a> fn(<fn(&'a ()) as Trait>::Assoc)) {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 2 previous errors
+

